src/operator/mshadow_op.h

```diff
@@ -99,22 +99,30 @@ struct reciprocal_grad {
 struct sigmoid {
   template<typename DType>
   MSHADOW_XINLINE static DType Map(DType a) {
-    return DType(DType(1.0f) / (DType(1.0f) + expf(-a)));
+    return DType(DType(1.0f) / (DType(1.0f) + std::expf(-a)));
   }
 };
+
+template<>
+MSHADOW_XINLINE double sigmoid::Map<double>(double a) {
+  return 1.0 / (1.0 + std::exp(-a));
```

Comment:
instead of specializing for each operator, I think it's better to make a templated function

template<typename DType>
mxnet_math::exp(DType a);

and specialize this.

Reply:
mxnet_math may not be a good namespace. Maybe mxnet_exp?
I'm open to suggestions

Reply:
I followed the example of lgamma (due to Valentin Flunkert, I think).

Would it work just to do template versions of the math.h functions?

I am thinking about these DType(...) expressions everywhere.

Reply:
`Would it work just to do template versions of the math.h functions?`

What does this mean? If you want to overload std::exp that's not a good idea.

Reply:
OK, I will be OOTO for a while, so I think this change has to wait (I may delete the PR).

Still, could you be more explicit how you would do it? Here is what I tried:
- Just added the template specializations for double for each XXX:Map
- Problem: Did not work for exp, expm1, log, ..., because now, you have a clash with
   the global names included via math.h
- At this point, I switched to including cmath, so it is std::exp, std::expm1, std::log.
   I agree this is not elegant, because lots of code has to change.

What would you suggest?

I am not even 100% why these name clashes really appear. For example, if I call
exp(a), where a is double, how can it confuse this with

struct exp { ... }

The latter does not have a constructor with double arg.

Sorry I am not so greatly familiar with C++. Maybe there is a much more elegant solution.
If you can suggest one, please tell me.

Reply:
We can implement our own set of math functions under the namespace mxnet::math. For example
```
namespace mxnet {
namespace math {
template<typename DType>
DType exp(DType value);

template<>
double exp(double value);

}
}